src/rapsV2/execute.ts

```diff
@@ -0,0 +1,121 @@
+  switch (type) {
+    case 'claimTransactionClaimableAction':
+      return () => claimTransactionClaimable(props);
+    default:
+      throw new RainbowError(`[rapsV2/execute]: T - unknown action type ${type}`);
+  }
+}
+
+// executes a single action in the rap
+// if the action executes a tx on-chain, it will return the nonce it used
+// if an error occurs, we return the error message
+export async function executeAction<T extends RapActionTypes>({
+  action,
+  wallet,
+  rap,
+  nonceToUse,
+  rapName,
+}: {
+  action: RapAction<T>;
+  wallet: Signer;
+  rap: Rap;
+  nonceToUse: number | undefined;
+  rapName: string;
+}): Promise<RapActionResponse> {
+  const { type, parameters } = action;
+  try {
+    const actionProps = {
+      wallet,
+      currentRap: rap,
+      parameters,
+      nonceToUse,
+    };
+    const { nonce, hash } = await getActionExecutableByType<T>(type, actionProps)();
+    return { nonce, errorMessage: null, hash };
+  } catch (error) {
+    logger.error(new RainbowError(`[rapsV2/execute]: ${rapName} - error execute action`), {
+      message: (error as Error)?.message,
+    });
+    return { nonce: null, errorMessage: String(error), hash: null };
+  }
+}
+
+function getRapFullName<T extends RapActionTypes>(actions: RapAction<T>[]) {
+  const actionTypes = actions.map(action => action.type);
+  return actionTypes.join(' + ');
+}
+
+const waitForNodeAck = async (hash: string, provider: Signer['provider']): Promise<void> => {
+  return new Promise(async resolve => {
+    const tx = await provider?.getTransaction(hash);
+    // This means the node is aware of the tx, we're good to go
+    if ((tx && tx.blockNumber === null) || (tx && tx?.blockNumber && tx?.blockNumber > 0)) {
+      resolve();
+    } else {
+      // Wait for 1 second and try again
+      await delay(1000);
+      return waitForNodeAck(hash, provider);
+    }
+  });
+};
+
+// goes through each action in the rap and executes it
+// if an action executes a tx on-chain, increment the nonceToUse for the next tx
+// if an action fails, it will return the error message
+const executeRap = async (wallet: Signer, rap: Rap): Promise<RapResponse> => {
+  const { actions } = rap;
+  const rapName = getRapFullName(rap.actions);
+  let nonceToUse: number | undefined;
+
+  while (actions.length) {
+    const action = actions.shift();
+
+    if (!action) break;
+
+    const { nonce, errorMessage, hash } = await executeAction({
+      action,
+      wallet,
+      rap,
+      nonceToUse,
+      rapName,
+    });
+
+    if (errorMessage) return { errorMessage };
+
+    if (typeof nonce === 'number') {
+      actions.length >= 1 && hash && (await waitForNodeAck(hash, wallet.provider));
+      nonceToUse = nonce + 1;
+    }
+  }
+
+  return { errorMessage: null };
+};
+
+export async function walletExecuteRap(wallet: Signer, rapParameters: RapParameters): Promise<RapResponse> {
+  const rap = await createRap(rapParameters);
+  return executeRap(wallet, rap);
+}
```

Comment:
i see a lot of this is a dupe from old raps, I understand there are types conflicts but there's no chance to avoid this?

Reply:
it's definitely avoidable, i felt that it was safer and more clear to keep v1 and v2 isolated. i can understand why that could be confusing when trying to tell which code is new and which is old.